src/lib.rs

```diff
@@ -0,0 +1,238 @@
+
+fn vec_multiref<S: 'static + N + T>() -> Vec<*const dyn T> {
+    vec![Box::into_raw(Box::new(S::new())); VEC_SIZE]
+}
+
+pub fn bench_normal_multiref_no_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_multiref::<SNoRead>();
+    time(|| {
+        for &e in &v {
+            assert_eq!(unsafe { (&*e).m() }, VAL_NOREAD);
+        }
+    });
+}
+
+pub fn bench_normal_multiref_with_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_multiref::<SWithRead>();
+    time(|| {
+        for &e in &v {
+            assert_eq!(unsafe { (&*e).m() }, VAL_WITHREAD);
+        }
+    });
+}
+
+fn vec_vtable<S: 'static + N + T>() -> Vec<*const u8> {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let mut v = Vec::with_capacity(VEC_SIZE);
+    // Since every instance of S will share the same vtable, it's OK for us to pull it out and
+    // reuse it. With the coerce_unsized feature turned on, we can do this in a marginally cleverer
+    // way, but the outcome is the same.
+    let vtable = {
+        let b: *const dyn T = Box::into_raw(Box::new(S::new()));
+        let (_, vtable) = unsafe { transmute::<_, (usize, usize)>(b) };
+        vtable
+    };
+    // We're going to lay out memory (on a 64-bit machine) as:
+    //   offset 0: vtable
+    //          8: S
+    let (layout, _) = Layout::new::<usize>().extend(Layout::new::<S>()).unwrap();
+    // The following assert ensure that the layout really is as we expect.
+    assert_eq!(layout.size(), size_of::<usize>() + size_of::<S>());
+    for _ in 0..VEC_SIZE {
+        let b = unsafe {
+            let b: *mut usize = alloc(layout) as *mut usize;
+            b.copy_from(&vtable, 1);
+            (b.add(1) as *mut S).copy_from(&S::new(), 1);
+            b as *const u8
+        };
+        v.push(b);
+    }
+    v
+}
+
+pub fn bench_alongside_no_read() {
+    let v = vec_vtable::<SNoRead>();
+    time(|| {
+        for &e in &v {
+            let vtable = unsafe { *(e as *const usize) };
+            let t_ptr = unsafe { (e as *const usize).add(1) };
+            let b: *const dyn T = unsafe { transmute((t_ptr, vtable)) };
+            assert_eq!(unsafe { (&*b).m() }, VAL_NOREAD);
+        }
+    });
+}
+
+pub fn bench_alongside_with_read() {
+    let v = vec_vtable::<SWithRead>();
+    time(|| {
+        for &e in &v {
+            let vtable = unsafe { *(e as *const usize) };
+            let t_ptr = unsafe { (e as *const usize).add(1) };
+            let b: *const dyn T = unsafe { transmute((t_ptr, vtable)) };
+            assert_eq!(unsafe { (&*b).m() }, VAL_WITHREAD);
+        }
+    });
+}
+
+fn vec_multiref_vtable<S: 'static + N + T>() -> Vec<*const u8> {
+    let mut v = Vec::with_capacity(VEC_SIZE);
+    let vtable = {
```

Comment:
Is the vtable extraction code the same as last time? Perhaps make it a function?